src/linker/Linker/TypeNameResolver.cs

```diff
@@ -31,13 +31,37 @@ public TypeReference ResolveTypeName (string typeNameString)
 				return ResolveTypeName (null, assemblyQualifiedTypeName);
 			}
 
-			foreach (var assemblyDefinition in _context.GetReferencedAssemblies ()) {
-				var foundType = ResolveTypeName (assemblyDefinition, parsedTypeName);
-				if (foundType != null)
-					return foundType;
-			}
+			// If parsedTypeName doesn't have an assembly name in it but it does have a namespace,
+			// search for the type in the calling object's assembly, mscorlib and SPCL, in that order.
+			AssemblyDefinition assemblyOfCaller = origin switch {
+				TypeReference tr => tr.Module.Assembly,
+				_ => origin.DeclaringType.Module.Assembly
+			};
+
+			if (assemblyOfCaller != null && TryResolveTypeName (assemblyOfCaller, parsedTypeName, out var typeRef))
+				return typeRef;
+
+			AssemblyDefinition mscorlibAssembly = _context.GetLoadedAssembly ("mscorlib");
+			if (mscorlibAssembly != null && TryResolveTypeName (mscorlibAssembly, parsedTypeName, out var typeRefFromMscorlib))
+				return typeRefFromMscorlib;
+
+			AssemblyDefinition spclAssembly = _context.GetLoadedAssembly (PlatformAssemblies.CoreLib);
+			if (spclAssembly != null && TryResolveTypeName (spclAssembly, parsedTypeName, out var typeRefFromSPCL))
+				return typeRefFromSPCL;
 
+			_context.LogWarning ($"Type '{typeNameString}' could not be found in the caller's assembly, 'mscorlib' nor 'System.Private.CoreLib'",
```

Comment:
We don't emit warnings like this because Type.GetType is often used in lightup (look for something using reflection and use it if it's there, fallback to something else if it isn't). This warning would be spurious.